File: workers/forecast_engine.py
"""Forecasting engine — weighted signal aggregation (V1).

Combines CRM pipeline, news intelligence, procurement notices and historical
patterns into a rolling 12-month demand forecast per skill. Trend and job-posting
weights are zeroed for the hackathon cut (those connectors ship in V1).
"""
from __future__ import annotations

import logging
import math
from datetime import date, datetime, timezone
from typing import Any

from supabase import Client, create_client

log = logging.getLogger(__name__)

CONVERGENCE_WINDOW_DAYS = 28
SIGNAL_RECENCY_WINDOW_DAYS = 90
MIN_PERSISTED_DEMAND = 0.1
SOURCE_CONFIDENCE_CAPS: dict[str, float] = {
    "crm_pipeline": 0.50,
    "procurement_notice": 0.75,
    "news_event": 0.55,
    "historical_pattern": 0.25,
    "trend_spike": 0.40,
    "job_posting": 0.55,
}


class ForecastEngine:
    DEFAULT_WEIGHTS: dict[str, float] = {
        "crm_pipeline": 0.35,
        "procurement_notice": 0.25,
        "historical_pattern": 0.15,
        "news_event": 0.05,
        "trend_spike": 0.10,
        "job_posting": 0.10,
    }

    def __init__(
        self,
        supabase_url: str,
        supabase_key: str,
        *,
        weights: dict[str, float] | None = None,
    ) -> None:
        self.db: Client = create_client(supabase_url, supabase_key)
        persisted = self._load_source_weights()
        self.weights = {**self.DEFAULT_WEIGHTS, **persisted, **(weights or {})}

    def _load_source_weights(self) -> dict[str, float]:
        try:
            rows = self.db.table("source_weights").select("source_key, weight").execute().data or []
        except Exception as exc:
            log.warning("failed to load source_weights, using defaults: %s", exc)
            return {}

        weights: dict[str, float] = {}
        for row in rows:
            key = row.get("source_key")
            value = row.get("weight")
            if not isinstance(key, str):
                continue
            try:
                weights[key] = float(value)
            except (TypeError, ValueError):
                continue
        return weights

    # ------------------------------------------------------------------ run --

    def run(self, months_ahead: int = 12) -> int:
        self._clear_future_forecasts()
        skills = self.db.table("skills").select("id, name").execute().data or []
        generated_at = datetime.now(timezone.utc).isoformat()
        rows_written = 0
        for skill in skills:
            forecasts = self.generate_forecast(
                skill["id"],
                months_ahead=months_ahead,
                generated_at=generated_at,
            )
            if not self._should_persist_skill(forecasts):
                continue
            for forecast in forecasts:
                self.db.table("forecasts").upsert(
                    forecast, on_conflict="forecast_month,skill_id"
                ).execute()
                rows_written += 1
        log.info("forecast engine wrote %d rows", rows_written)
        return rows_written

    # ---------------------------------------------------------- generation --

    def generate_forecast(
        self,
        skill_id: str,
        months_ahead: int = 12,
        *,
        generated_at: str | None = None,
    ) -> list[dict]:
        results: list[dict] = []
        stamp = generated_at or datetime.now(timezone.utc).isoformat()

        for offset in range(1, months_ahead + 1):
            target_month = _add_months(_start_of_month(date.today()), offset)

            crm = self._score_crm_pipeline(skill_id, target_month)
            procurement = self._score_procurement(skill_id, target_month)
            news = self._score_news_events(skill_id, target_month)
            trend = self._score_trends(skill_id, target_month)
            postings = self._score_job_postings(skill_id, target_month)
            has_current_signal = (
                crm > 0 or procurement > 0 or news > 0 or trend > 0 or postings > 0
            )
            historical = (
                self._score_historical_pattern(skill_id, target_month)
                if has_current_signal
                else 0.0
            )

            raw_demand = (
                crm * self.weights["crm_pipeline"]
                + procurement * self.weights["procurement_notice"]
                + historical * self.weights["historical_pattern"]
                + news * self.weights["news_event"]
                + trend * self.weights["trend_spike"]
                + postings * self.weights["job_posting"]
            )

            source_confidences = {
                "crm_pipeline": self._effective_source_confidence(
                    "crm_pipeline", crm
                ),
                "procurement_notice": self._effective_source_confidence(
                    "procurement_notice", procurement
                ),
                "news_event": self._effective_source_confidence("news_event", news),
                "trend_spike": self._effective_source_confidence(
                    "trend_spike", trend
                ),
                "job_posting": self._effective_source_confidence(
                    "job_posting", postings
                ),
                "historical_pattern": self._effective_source_confidence(
                    "historical_pattern", historical
                ),
            }
            confidence = self._combine_source_confidence(source_confidences)
            converging = sum(
                1
                for source, value in source_confidences.items()
                if value > 0 and source != "historical_pattern"
            ) >= 2
            supply = self._get_supply(skill_id, target_month)

            results.append(
                {
                    "generated_at": stamp,
                    "skill_id": skill_id,
                    "forecast_month": target_month.isoformat(),
                    "predicted_demand": round(raw_demand, 2),
                    "current_supply": supply,
                    "gap": round(raw_demand - supply, 2),
                    "confidence": round(confidence, 2),
                    "contributing_signals": self._contributing_signal_ids(
                        skill_id, target_month
                    ),
                    "notes": self._explain(
                        skill_id,
                        crm,
                        procurement,
                        news,
                        historical,
                        trend,
                        postings,
                        converging,
                    ),
                }
            )
        return results

    # ---------------------------------------------------- signal scorers ----

    def _clear_future_forecasts(self) -> None:
        start = _start_of_month(date.today()).isoformat()
        self.db.table("forecasts").delete().gte("forecast_month", start).execute()

    @staticmethod
    def _should_persist_skill(forecasts: list[dict]) -> bool:
        return any(
            float(forecast.get("predicted_demand") or 0) >= MIN_PERSISTED_DEMAND
            for forecast in forecasts
        )

    @staticmethod
    def _effective_source_confidence(source: str, score: float) -> float:
        if score <= 0:
            return 0.0
        cap = SOURCE_CONFIDENCE_CAPS[source]
        return min(cap, 1 - math.exp(-score))

    @staticmethod
    def _combine_source_confidence(source_confidences: dict[str, float]) -> float:
        active = {
            source: value for source, value in source_confidences.items() if value > 0
        }
        if not active:
            return 0.0

        combined = 1.0 - math.prod(1.0 - value for value in active.values())
        supporting = [
            source for source in active.keys() if source != "historical_pattern"
        ]

        if not supporting:
            return round(min(combined, 0.25), 2)
        if len(supporting) == 1:
            return round(min(combined, SOURCE_CONFIDENCE_CAPS[supporting[0]]), 2)
        return round(min(combined + 0.1, 0.95), 2)

    def _score_crm_pipeline(self, skill_id: str, target_month: date) -> float:
        deals = (
            self.db.table("deal_profiles")
            .select("quantity, deals(expected_start, probability, status)")
            .eq("skill_id", skill_id)
            .execute()
            .data
            or []
        )
        score = 0.0
        current_month = _start_of_month(date.today())
        target_offset = _month_diff(target_month, current_month)
        for row in deals:
            deal = row.get("deals") or {}
            if deal.get("status") in ("won", "lost"):
                continue
            start = deal.get("expected_start")
            probability = deal.get("probability", 0.5)
            if not start:
                continue
            start_month = _start_of_month(date.fromisoformat(start))
            months_from_now = _month_diff(start_month, current_month)
            proximity = max(0.0, 1 - abs(months_from_now - target_offset) / 12)
            score += row["quantity"] * probability * proximity
        return score

    def _score_procurement(self, skill_id: str, target_month: date) -> float:
        return self._score_signal_sources(
            skill_id,
            sources=("ted_procurement",),
            target_month=target_month,
            window_days=SIGNAL_RECENCY_WINDOW_DAYS,
        )

    def _score_news_events(self, skill_id: str, target_month: date) -> float:
        return self._score_signal_sources(
            skill_id,
            sources=("news_intelligence", "news"),
            target_month=target_month,
            window_days=SIGNAL_RECENCY_WINDOW_DAYS,
        )

    def _score_trends(self, skill_id: str, target_month: date) -> float:
        return self._score_signal_sources(
            skill_id,
            sources=("google_trends",),
            target_month=target_month,
            window_days=SIGNAL_RECENCY_WINDOW_DAYS,
        )

    def _score_job_postings(self, skill_id: str, target_month: date) -> float:
        return self._score_signal_sources(
            skill_id,
            sources=("ats_greenhouse", "ats_lever"),
            target_month=target_month,
            window_days=SIGNAL_RECENCY_WINDOW_DAYS,
        )

    def _score_signal_sources(
        self,
        skill_id: str,
        *,
        sources: tuple[str, ...],
        target_month: date,
        window_days: int,
    ) -> float:
        rows = (
            self.db.table("signal_skills")
            .select("confidence, signals(detected_at, source)")
            .eq("skill_id", skill_id)
            .execute()
            .data
            or []
        )
        score = 0.0
        allowed = set(sources)
        for row in rows:
            sig = row.get("signals") or {}
            if sig.get("source") not in allowed:
                continue
            detected = _parse_detected_at(sig.get("detected_at"))
            if detected is None:
                continue
            days_age = abs((target_month - detected).days)
            if days_age > window_days:
                continue
            decay = max(0.0, 1.0 - days_age / window_days)
            score += row["confidence"] * decay
        return score

    def _score_historical_pattern(self, skill_id: str, target_month: date) -> float:
        rows = (
            self.db.table("project_skills")
            .select("headcount, projects(started_at)")
            .eq("skill_id", skill_id)
            .execute()
            .data
            or []
        )
        if not rows:
            return 0.0
        month_counts: dict[int, int] = {}
        for row in rows:
            started_at = (row.get("projects") or {}).get("started_at")
            if not started_at:
                continue
            month = date.fromisoformat(started_at).month
            month_counts[month] = month_counts.get(month, 0) + row["headcount"]
        total = sum(month_counts.values()) or 1
        return month_counts.get(target_month.month, 0) / total

    def _get_supply(self, skill_id: str, target_month: date) -> int:
        rows = (
            self.db.table("consultant_skills")
            .select("consultants(current_status, available_from)")
            .eq("skill_id", skill_id)
            .execute()
            .data
            or []
        )
        available = 0
        for row in rows:
            consultant = row.get("consultants") or {}
            if consultant.get("current_status") == "on_mission":
                continue
            avail_from = consultant.get("available_from")
            if avail_from and date.fromisoformat(avail_from) > target_month:
                continue
            available += 1
        return available

    def _contributing_signal_ids(
        self, skill_id: str, target_month: date
    ) -> list[str]:
        rows = (
            self.db.table("signal_skills")
            .select("signal_id, signals(detected_at)")
            .eq("skill_id", skill_id)
            .execute()
            .data
            or []
        )
        ids: list[str] = []
        for row in rows:
            detected = _parse_detected_at((row.get("signals") or {}).get("detected_at"))
            if detected is None:
                continue
            if abs((target_month - detected).days) <= CONVERGENCE_WINDOW_DAYS:
                ids.append(row["signal_id"])
        return sorted(set(ids))

    def _explain(
        self,
        skill_id: str,
        crm: float,
        procurement: float,
        news: float,
        historical: float,
        trend: float,
        postings: float,
        converging: bool,
    ) -> str:
        skill = (
            self.db.table("skills")
            .select("name")
            .eq("id", skill_id)
            .single()
            .execute()
            .data
            or {}
        )
        name = skill.get("name", "this skill")
        parts: list[str] = []
        if crm > 0:
            parts.append(f"Pipeline deals request {name}")
        if procurement > 0:
            parts.append(f"Recent procurement notices mention {name}")
        if news > 0:
            parts.append(f"Belgian news coverage signals demand for {name}")
        if historical > 0:
            parts.append("Seasonality: demand typically rises this quarter")
        if trend > 0:
            parts.append(f"Trend signals show growing attention around {name}")
        if postings > 0:
            parts.append(f"Live job postings point to hiring pressure for {name}")
        if converging:
            parts.append("Multiple independent sources agree — confidence boosted")
        return ". ".join(parts) + "." if parts else "No strong signals detected."


def _parse_detected_at(value: Any) -> date | None:
    """Parse a Supabase `detected_at` timestamp into a date (best-effort)."""
    if value is None:
        return None
    if isinstance(value, date) and not isinstance(value, datetime):
        return value
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return None
        if text.endswith("Z"):
            text = text[:-1] + "+00:00"
        try:
            return datetime.fromisoformat(text).astimezone(timezone.utc).date()
        except ValueError:
            try:
                return date.fromisoformat(text[:10])
            except ValueError:
                return None
    return None


def _start_of_month(value: date) -> date:
    return value.replace(day=1)


def _add_months(value: date, months: int) -> date:
    month_index = value.month - 1 + months
    year = value.year + month_index // 12
    month = month_index % 12 + 1
    return date(year, month, 1)


def _month_diff(a: date, b: date) -> int:
    return (a.year - b.year) * 12 + (a.month - b.month)

File: lib/server/forecast-engine.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import type { ForecastFactorKey, SignalSource } from "@/lib/types";

const FORECAST_WEEKS = 12;
const CONVERGENCE_WINDOW_DAYS = 28;
const SIGNAL_RECENCY_WINDOW_DAYS = 90;
const CONFIDENCE_DIVISOR = 4;
const UPSERT_BATCH_SIZE = 500;
const MS_PER_DAY = 24 * 60 * 60 * 1000;
const MS_PER_YEAR = 365.25 * MS_PER_DAY;

export const HALF_LIFE_YEARS = 2;
export const BENCH_PRESSURE_GAIN = 0.5;
export const MIN_HISTORY_WEIGHTED_STARTS = 6;
export const BASELINE_WINDOW_MONTHS = 24;
export const BENCH_DURATION_WINDOW_MONTHS = 18;
export const HISTORY_DAMPEN_WITHOUT_CURRENT_SIGNAL = 0.5;

export interface HistoricalAggregate {
  seasonalIndex: Record<number, number>;
  weightedMonthly: Record<number, number>;
  baselineWeekly: number;
  tightness: number;
  skillMedianDuration: number | null;
}

export interface HistoricalDataset {
  aggregates: Map<string, HistoricalAggregate>;
  globalMedianDuration: number | null;
  startsBySkill: Map<string, Date[]>;
}

interface SkillRow {
  id: string;
  name: string;
}

interface DealRelationship {
  expected_start: string | null;
  probability: number | null;
  status: string | null;
}

interface DealProfileRow {
  skill_id: string;
  quantity: number | null;
  deals: DealRelationship | DealRelationship[] | null;
}

interface SignalRelationship {
  detected_at: string | null;
  source: SignalSource | null;
}

interface SignalSkillRow {
  skill_id: string;
  signal_id: string;
  confidence: number | null;
  signals: SignalRelationship | SignalRelationship[] | null;
}

interface ProjectRelationship {
  started_at: string | null;
  source_kind: string | null;
}

interface ProjectSkillRow {
  skill_id: string;
  projects: ProjectRelationship | ProjectRelationship[] | null;
}

interface BenchHistoryRelationship {
  duration_days: number | null;
  bench_ended_at: string | null;
}

interface BenchHistorySkillRow {
  skill_id: string;
  consultant_bench_history:
    | BenchHistoryRelationship
    | BenchHistoryRelationship[]
    | null;
}

interface ConsultantRelationship {
  current_status: string | null;
  available_from: string | null;
}

interface ConsultantSkillRow {
  skill_id: string;
  consultants: ConsultantRelationship | ConsultantRelationship[] | null;
}

interface ForecastInsertRow {
  generated_at: string;
  forecast_week: string;
  skill_id: string;
  predicted_demand: number;
  current_supply: number;
  gap: number;
  confidence: number;
  contributing_signals: string[];
  notes: string;
}

export interface ForecastRecalculationResult {
  generated_at: string;
  skills_processed: number;
  forecasts_written: number;
}

function asSingle<T>(value: T | T[] | null | undefined): T | null {
  if (Array.isArray(value)) return value[0] ?? null;
  return value ?? null;
}

function startOfUtcDay(date: Date): Date {
  return new Date(
    Date.UTC(date.getUTCFullYear(), date.getUTCMonth(), date.getUTCDate()),
  );
}

function addDays(date: Date, days: number): Date {
  return new Date(date.getTime() + days * MS_PER_DAY);
}

function parseDateValue(value: string | null | undefined): Date | null {
  if (!value) return null;

  const text = value.trim();
  if (!text) return null;

  const parsed = /^\d{4}-\d{2}-\d{2}$/.test(text)
    ? new Date(`${text}T00:00:00.000Z`)
    : new Date(text);

  if (Number.isNaN(parsed.getTime())) return null;
  return startOfUtcDay(parsed);
}

function isoDate(date: Date): string {
  return date.toISOString().slice(0, 10);
}

function daysBetween(a: Date, b: Date): number {
  return Math.round((a.getTime() - b.getTime()) / MS_PER_DAY);
}

function roundTo(value: number, digits = 2): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function groupBySkill<T extends { skill_id: string }>(rows: T[]): Map<string, T[]> {
  const grouped = new Map<string, T[]>();
  for (const row of rows) {
    const bucket = grouped.get(row.skill_id) ?? [];
    bucket.push(row);
    grouped.set(row.skill_id, bucket);
  }
  return grouped;
}

function scoreSignalSources(
  rows: SignalSkillRow[],
  targetWeek: Date,
  sources: SignalSource[],
): number {
  let score = 0;
  const allowed = new Set(sources);

  for (const row of rows) {
    const signal = asSingle(row.signals);
    if (!signal?.source || !allowed.has(signal.source)) continue;

    const detectedAt = parseDateValue(signal.detected_at);
    if (!detectedAt) continue;

    const daysAge = Math.abs(daysBetween(targetWeek, detectedAt));
    if (daysAge > SIGNAL_RECENCY_WINDOW_DAYS) continue;

    const decay = Math.max(0, 1 - daysAge / SIGNAL_RECENCY_WINDOW_DAYS);
    score += (row.confidence ?? 0) * decay;
  }

  return score;
}

function scoreCrmPipeline(rows: DealProfileRow[], today: Date, targetWeek: Date): number {
  let score = 0;
  const targetWeeksFromNow = daysBetween(targetWeek, today) / 7;

  for (const row of rows) {
    const deal = asSingle(row.deals);
    if (!deal?.expected_start) continue;
    if (deal.status && ["won", "lost"].includes(deal.status)) continue;

    const startDate = parseDateValue(deal.expected_start);
    if (!startDate) continue;

    const weeksFromNow = daysBetween(startDate, today) / 7;
    const proximity = Math.max(
      0,
      1 - Math.abs(weeksFromNow - targetWeeksFromNow) / 12,
    );

    score += (row.quantity ?? 0) * (deal.probability ?? 0.5) * proximity;
  }

  return score;
}

function median(values: number[]): number {
  if (!values.length) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 0
    ? (sorted[mid - 1] + sorted[mid]) / 2
    : sorted[mid];
}

function emptyMonthlyMap(): Record<number, number> {
  return Object.fromEntries(
    Array.from({ length: 12 }, (_, i) => [i + 1, 0]),
  ) as Record<number, number>;
}

function flatSeasonalIndex(): Record<number, number> {
  return Object.fromEntries(
    Array.from({ length: 12 }, (_, i) => [i + 1, 1]),
  ) as Record<number, number>;
}

function buildHistoricalAggregate(
  starts: Date[],
  durations: number[],
  today: Date,
  globalMedianDuration: number | null,
): HistoricalAggregate {
  const aggregate: HistoricalAggregate = {
    seasonalIndex: flatSeasonalIndex(),
    weightedMonthly: emptyMonthlyMap(),
    baselineWeekly: 0,
    tightness: 0,
    skillMedianDuration: null,
  };

  if (!starts.length && !durations.length) return aggregate;

  const baselineCutoff = addDays(today, -BASELINE_WINDOW_MONTHS * 30);
  let totalWeighted = 0;
  let baselineWeighted = 0;

  for (const start of starts) {
    const yearsAgo = (today.getTime() - start.getTime()) / MS_PER_YEAR;
    if (yearsAgo < 0) continue;
    const weight = Math.exp(-yearsAgo / HALF_LIFE_YEARS);
    const month = start.getUTCMonth() + 1;
    aggregate.weightedMonthly[month] += weight;
    totalWeighted += weight;
    if (start.getTime() >= baselineCutoff.getTime()) {
      baselineWeighted += weight;
    }
  }

  if (totalWeighted >= MIN_HISTORY_WEIGHTED_STARTS) {
    const avg = totalWeighted / 12;
    if (avg > 0) {
      aggregate.seasonalIndex = Object.fromEntries(
        Array.from({ length: 12 }, (_, i) => [
          i + 1,
          aggregate.weightedMonthly[i + 1] / avg,
        ]),
      ) as Record<number, number>;
    }
  }

  aggregate.baselineWeekly = baselineWeighted / 52;

  if (durations.length) {
    const skillMedian = median(durations);
    aggregate.skillMedianDuration = skillMedian;
    if (globalMedianDuration !== null && globalMedianDuration > 0) {
      const tightness = (globalMedianDuration - skillMedian) / globalMedianDuration;
      aggregate.tightness = Math.max(-1, Math.min(1, tightness));
    }
  }

  return aggregate;
}

export async function loadHistoricalDataset(
  supabase: SupabaseClient,
  referenceDate: Date = startOfUtcDay(new Date()),
): Promise<HistoricalDataset> {
  const [projectSkillsResult, benchSkillsResult] = await Promise.all([
    supabase
      .from("project_skills")
      .select("skill_id, projects(started_at, source_kind)"),
    supabase
      .from("consultant_bench_history_skills")
      .select(
        "skill_id, consultant_bench_history(duration_days, bench_ended_at)",
      ),
  ]);

  if (projectSkillsResult.error) throw projectSkillsResult.error;
  if (benchSkillsResult.error) throw benchSkillsResult.error;

  const startsBySkill = new Map<string, Date[]>();
  for (const row of (projectSkillsResult.data ?? []) as ProjectSkillRow[]) {
    const project = asSingle(row.projects);
    if (project?.source_kind !== "history") continue;
    const started = parseDateValue(project.started_at);
    if (!started) continue;
    const bucket = startsBySkill.get(row.skill_id) ?? [];
    bucket.push(started);
    startsBySkill.set(row.skill_id, bucket);
  }

  const benchCutoff = addDays(referenceDate, -BENCH_DURATION_WINDOW_MONTHS * 30);
  const durationsBySkill = new Map<string, number[]>();
  const allDurations: number[] = [];
  for (const row of (benchSkillsResult.data ?? []) as BenchHistorySkillRow[]) {
    const spell = asSingle(row.consultant_bench_history);
    if (!spell) continue;
    const duration = spell.duration_days;
    const ended = parseDateValue(spell.bench_ended_at);
    if (duration === null || duration === undefined || !ended) continue;
    if (ended.getTime() < benchCutoff.getTime()) continue;
    const bucket = durationsBySkill.get(row.skill_id) ?? [];
    bucket.push(duration);
    durationsBySkill.set(row.skill_id, bucket);
    allDurations.push(duration);
  }

  const globalMedianDuration = allDurations.length ? median(allDurations) : null;

  const aggregates = new Map<string, HistoricalAggregate>();
  const skillIds = new Set<string>([
    ...startsBySkill.keys(),
    ...durationsBySkill.keys(),
  ]);
  for (const skillId of skillIds) {
    aggregates.set(
      skillId,
      buildHistoricalAggregate(
        startsBySkill.get(skillId) ?? [],
        durationsBySkill.get(skillId) ?? [],
        referenceDate,
        globalMedianDuration,
      ),
    );
  }

  return { aggregates, globalMedianDuration, startsBySkill };
}

function scoreHistoricalPattern(
  aggregate: HistoricalAggregate | undefined,
  targetWeek: Date,
): number {
  if (!aggregate || aggregate.baselineWeekly <= 0) return 0;
  const month = targetWeek.getUTCMonth() + 1;
  const pressureMult = 1 + BENCH_PRESSURE_GAIN * Math.max(0, aggregate.tightness);
  return aggregate.baselineWeekly * (aggregate.seasonalIndex[month] ?? 1) * pressureMult;
}

function getSupply(rows: ConsultantSkillRow[], targetWeek: Date): number {
  let available = 0;

  for (const row of rows) {
    const consultant = asSingle(row.consultants);
    if (!consultant) continue;
    if (consultant.current_status === "on_mission") continue;

    const availableFrom = parseDateValue(consultant.available_from);
    if (availableFrom && availableFrom.getTime() > targetWeek.getTime()) continue;

    available += 1;
  }

  return available;
}

function contributingSignalIds(rows: SignalSkillRow[], targetWeek: Date): string[] {
  const ids = new Set<string>();

  for (const row of rows) {
    const signal = asSingle(row.signals);
    const detectedAt = parseDateValue(signal?.detected_at);
    if (!detectedAt) continue;

    if (Math.abs(daysBetween(targetWeek, detectedAt)) <= CONVERGENCE_WINDOW_DAYS) {
      ids.add(row.signal_id);
    }
  }

  return Array.from(ids);
}

function convergentSources(rows: SignalSkillRow[], targetWeek: Date): Set<string> {
  const sources = new Set<string>();

  for (const row of rows) {
    const signal = asSingle(row.signals);
    if (!signal?.source) continue;

    const detectedAt = parseDateValue(signal.detected_at);
    if (!detectedAt) continue;

    if (Math.abs(daysBetween(targetWeek, detectedAt)) <= CONVERGENCE_WINDOW_DAYS) {
      sources.add(signal.source);
    }
  }

  return sources;
}

function explainForecast(
  skillName: string,
  scores: Record<ForecastFactorKey, number>,
  converging: boolean,
  aggregate: HistoricalAggregate | undefined,
  targetWeek: Date,
): string {
  const parts: string[] = [];

  if (scores.crm_pipeline > 0) {
    parts.push(`Pipeline deals request ${skillName}`);
  }
  if (scores.procurement_notice > 0) {
    parts.push(`Recent procurement notices mention ${skillName}`);
  }
  if (scores.news_event > 0) {
    parts.push(`Belgian news coverage signals demand for ${skillName}`);
  }
  if (scores.historical_pattern > 0) {
    if (aggregate) {
      const month = targetWeek.getUTCMonth() + 1;
      const index = aggregate.seasonalIndex[month] ?? 1;
      if (index > 1.2) {
        parts.push(`Historically busier than average this month for ${skillName}`);
      } else if (index < 0.8) {
        parts.push(`Historically quieter than average this month for ${skillName}`);
      } else {
        parts.push(`Baseline historical demand for ${skillName}`);
      }
      if (aggregate.tightness > 0.4) {
        parts.push(`Bench clears fast for ${skillName} — market is tight`);
      }
    } else {
      parts.push(`Historical baseline for ${skillName}`);
    }
  }
  if (scores.trend_spike > 0) {
    parts.push(`Trend signals show growing attention around ${skillName}`);
  }
  if (scores.job_posting > 0) {
    parts.push(`Live job postings point to hiring pressure for ${skillName}`);
  }
  if (converging) {
    parts.push("Multiple independent sources agree — confidence boosted");
  }

  return parts.length ? `${parts.join(". ")}.` : "No strong signals detected.";
}

export async function generateAndPersistForecasts(
  supabase: SupabaseClient,
  weights: Record<ForecastFactorKey, number>,
  weeksAhead = FORECAST_WEEKS,
): Promise<ForecastRecalculationResult> {
  const today = startOfUtcDay(new Date());

  const [
    skillsResult,
    dealProfilesResult,
    signalSkillsResult,
    consultantSkillsResult,
    historicalDataset,
  ] = await Promise.all([
    supabase.from("skills").select("id, name").order("name", { ascending: true }),
    supabase
      .from("deal_profiles")
      .select("skill_id, quantity, deals(expected_start, probability, status)"),
    supabase
      .from("signal_skills")
      .select("skill_id, signal_id, confidence, signals(detected_at, source)"),
    supabase
      .from("consultant_skills")
      .select("skill_id, consultants(current_status, available_from)"),
    loadHistoricalDataset(supabase, today),
  ]);

  if (skillsResult.error) throw skillsResult.error;
  if (dealProfilesResult.error) throw dealProfilesResult.error;
  if (signalSkillsResult.error) throw signalSkillsResult.error;
  if (consultantSkillsResult.error) throw consultantSkillsResult.error;

  const skills = (skillsResult.data ?? []) as SkillRow[];
  const dealsBySkill = groupBySkill((dealProfilesResult.data ?? []) as DealProfileRow[]);
  const signalsBySkill = groupBySkill(
    (signalSkillsResult.data ?? []) as SignalSkillRow[],
  );
  const consultantsBySkill = groupBySkill(
    (consultantSkillsResult.data ?? []) as ConsultantSkillRow[],
  );

  const generatedAt = new Date().toISOString();
  const forecastRows: ForecastInsertRow[] = [];

  for (const skill of skills) {
    const dealRows = dealsBySkill.get(skill.id) ?? [];
    const signalRows = signalsBySkill.get(skill.id) ?? [];
    const consultantRows = consultantsBySkill.get(skill.id) ?? [];
    const aggregate = historicalDataset.aggregates.get(skill.id);

    for (let offset = 1; offset <= weeksAhead; offset += 1) {
      const targetWeek = addDays(today, offset * 7);

      const crm = scoreCrmPipeline(dealRows, today, targetWeek);
      const procurement = scoreSignalSources(signalRows, targetWeek, [
        "ted_procurement",
      ]);
      const news = scoreSignalSources(signalRows, targetWeek, [
        "news",
        "news_intelligence",
      ]);
      const trend = scoreSignalSources(signalRows, targetWeek, ["google_trends"]);
      const postings = scoreSignalSources(signalRows, targetWeek, [
        "ats_greenhouse",
        "ats_lever",
      ]);
      const hasCurrentSignal =
        crm > 0 || procurement > 0 || news > 0 || trend > 0 || postings > 0;
      let historical = scoreHistoricalPattern(aggregate, targetWeek);
      if (!hasCurrentSignal) {
        historical *= HISTORY_DAMPEN_WITHOUT_CURRENT_SIGNAL;
      }

      const scores: Record<ForecastFactorKey, number> = {
        crm_pipeline: crm,
        procurement_notice: procurement,
        historical_pattern: historical,
        news_event: news,
        trend_spike: trend,
        job_posting: postings,
      };

      const rawDemand =
        scores.crm_pipeline * weights.crm_pipeline +
        scores.procurement_notice * weights.procurement_notice +
        scores.historical_pattern * weights.historical_pattern +
        scores.news_event * weights.news_event +
        scores.trend_spike * weights.trend_spike +
        scores.job_posting * weights.job_posting;

      const activeSignals = Object.values(scores).filter((score) => score > 0).length;
      const baseConfidence = Math.min(activeSignals / CONFIDENCE_DIVISOR, 1);
      const converging = convergentSources(signalRows, targetWeek).size >= 2;
      const confidence = Math.min(
        baseConfidence + (converging ? 0.2 : 0),
        1,
      );

      const supply = getSupply(consultantRows, targetWeek);

      forecastRows.push({
        generated_at: generatedAt,
        forecast_week: isoDate(targetWeek),
        skill_id: skill.id,
        predicted_demand: roundTo(rawDemand),
        current_supply: supply,
        gap: roundTo(rawDemand - supply),
        confidence: roundTo(confidence),
        contributing_signals: contributingSignalIds(signalRows, targetWeek),
        notes: explainForecast(skill.name, scores, converging, aggregate, targetWeek),
      });
    }
  }

  for (let index = 0; index < forecastRows.length; index += UPSERT_BATCH_SIZE) {
    const batch = forecastRows.slice(index, index + UPSERT_BATCH_SIZE);
    const { error } = await supabase
      .from("forecasts")
      .upsert(batch, { onConflict: "forecast_week,skill_id" });

    if (error) throw error;
  }

  return {
    generated_at: generatedAt,
    skills_processed: skills.length,
    forecasts_written: forecastRows.length,
  };
}
